js2-mode.el

```diff
@@ -7150,67 +7194,91 @@ are ignored."
-                          (js2-wrapper-function-p parent))
-                     (setq inited (if (memq node (js2-function-node-params parent)) ?P t)))
-
-                    ((js2-for-in-node-p parent)
-                     (if (eq node (js2-for-in-node-iterator parent))
-                         (setq inited t used nil)))
-
-                    ((js2-function-node-p parent)
-                     (setq inited (if (memq node (js2-function-node-params parent)) ?P t)
-                           used nil)))
-
-                   (unless used
-                     (let ((grandparent (js2-node-parent parent)))
-                       (when grandparent
-                         (setq used (js2-return-node-p grandparent)))))
-
-                   (js2--add-or-update-symbol node inited used vars))))))))
+                    ((js2-var-init-node-p granparent)
+                     (when (null var-init-node)
+                       (setq var-init-node granparent)))
+                    ((js2-assign-node-p granparent)
+                     (when (null assign-node)
+                       (setq assign-node granparent))))
+                   (setq granparent (js2-node-parent granparent)))
+
+                 ;; If we are within a var-init-node, determine if the name is
+                 ;; declared and initialized
+                 (when var-init-node
+                   (let ((target (js2-var-init-node-target var-init-node)))
+                     (setq declared (memq node (js2--collect-target-symbols target nil)))
+                     ;; Is there an initializer for the declared variable?
+                     (when (js2-var-init-node-initializer var-init-node)
+                       (setq assigned declared)
+                       ;; Determine if the name is actually a literal object key
+                       ;; that we shall ignore later
+                       (when (and (not declared)
+                                  (js2-object-prop-node-p parent)
+                                  (eq node (js2-object-prop-node-left parent)))
+                         (setq object-key t)))
+                     ;; Maybe this is a for loop and the variable is one of its iterators?
+                     (unless assigned
+                       (let* ((gp (js2-node-parent parent))
+                              (ggp (if gp (js2-node-parent gp))))
+                         (when (and ggp (js2-for-in-node-p ggp))
+                           (setq assigned
+                                 (memq node
+                                       (cl-loop for kid in (js2-var-decl-node-kids
+                                                            (js2-for-in-node-iterator ggp))
+                                                with syms = '()
+                                                do (setq syms
+                                                         (append syms
+                                                                 (js2--collect-target-symbols
```

Comment:
This looks a little too wide (we generally try to keep it within the default 80 columns).

Otherwise, I'm happy to merge.